### PR TITLE
dataflow/operators/core: add `.partition()` for `StreamCore`

### PR DESCRIPTION
**Pull request by petrosagg**

### CODE REVIEW

File: timely/src/dataflow/operators/core/partition.rs [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+    ///                          .partition(3, |x| (x % 3, x));
+    ///
+    ///     for (idx, stream) in streams.into_iter().enumerate() {
+    ///         stream
+    ///             .container::<Vec<_>>()
+    ///             .inspect(move |x| println!("seen {idx}: {x:?}"));
+    ///     }
+    /// });
+    /// ```
+    fn partition<C2, D2, F>(&self, parts: u64, route: F) -> Vec<StreamCore<G, C2>>
```
Comment (antiguru) [inline timely/src/dataflow/operators/core/partition.rs:30]: Consider taking a container builder here instead!

Comment (petrosagg) [reply timely/src/dataflow/operators/core/partition.rs]: Good point! Changed to a container builder argument

**Comment by frankmcsherry** [comment]
Looks good to me, thanks!

File: timely/src/dataflow/operators/core/partition.rs [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+    ///         stream
+    ///             .container::<Vec<_>>()
+    ///             .inspect(move |x| println!("seen {idx}: {x:?}"));
+    ///     }
+    /// });
+    /// ```
+    fn partition<CB, D2, F>(&self, parts: u64, route: F) -> Vec<StreamCore<G, CB::Container>>
+    where
+        CB: ContainerBuilder,
+        CB::Container: SizableContainer + PushInto<D2> + Data,
```
Comment (antiguru) [inline timely/src/dataflow/operators/core/partition.rs:33]: The `SizableContainer` requirement here is not needed, right? That's a requirement that'll be imposed by the container builder, but the implementation should only require the remaining traits for the container.

Comment (antiguru) [reply timely/src/dataflow/operators/core/partition.rs:33]: Fixed in #612.

**Pull request merged**